internal/controllers/machinedeployment/machinedeployment_sync.go

```diff
@@ -86,173 +89,266 @@ func (r *Reconciler) getAllMachineSetsAndSyncRevision(ctx context.Context, d *cl
 	return newMS, allOldMSs, nil
 }
 
-// Returns a machine set that matches the intent of the given deployment. Returns nil if the new machine set doesn't exist yet.
-// 1. Get existing new MS (the MS that the given deployment targets, whose machine template is the same as deployment's).
-// 2. If there's existing new MS, update its revision number if it's smaller than (maxOldRevision + 1), where maxOldRevision is the max revision number among all old MSes.
-// 3. If there's no existing new MS and createIfNotExisted is true, create one with appropriate revision number (maxOldRevision + 1) and replicas.
-// Note that the machine-template-hash will be added to adopted MSes and machines.
-func (r *Reconciler) getNewMachineSet(ctx context.Context, d *clusterv1.MachineDeployment, msList, oldMSs []*clusterv1.MachineSet, createIfNotExisted bool) (*clusterv1.MachineSet, error) {
-	log := ctrl.LoggerFrom(ctx)
-
-	existingNewMS := mdutil.FindNewMachineSet(d, msList)
-
-	// Calculate the max revision number among all old MSes
-	maxOldRevision := mdutil.MaxRevision(oldMSs, log)
-
-	// Calculate revision number for this new machine set
-	newRevision := strconv.FormatInt(maxOldRevision+1, 10)
-
-	// Latest machine set exists. We need to sync its annotations (includes copying all but
-	// annotationsToSkip from the parent deployment, and update revision, desiredReplicas,
-	// and maxReplicas) and also update the revision annotation in the deployment with the
-	// latest revision.
-	if existingNewMS != nil {
-		msCopy := existingNewMS.DeepCopy()
-		patchHelper, err := patch.NewHelper(msCopy, r.Client)
+// Returns a MachineSet that matches the intent of the given MachineDeployment.
+// If there does not exist such a MachineSet and createIfNotExisted is true, create a new MachineSet.
+// If there is already such a MachineSet, update it to propagate in-place mutable fields from the MachineDeployment.
+func (r *Reconciler) getNewMachineSet(ctx context.Context, d *clusterv1.MachineDeployment, msList, oldMSs []*clusterv1.MachineSet, createIfNotExists bool) (*clusterv1.MachineSet, error) {
```

Comment:
If they do not already exists, we should add unit test for this func verifying at least
- Creates new MS (first MS)
- Do not create new MS (first MS)
- Update an existing MS (in place changes)
- Create a new MS (rollout changes)

Reply:
This functionality is tested/covered under with `TestMachineDeploymentReconciler`. 
I extended the tests to also properly cover the in-place propagation along with all other cases.

The tests for `getNewMachineSet` would largely be duplicates of the test cases covered under `TestMachineDeploymentReconciler` and would make the tests hard to maintain. No strong opinions if we want to add the tests. 🙂